subxt/src/custom_types/mod.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! Types associated with accessing custom types
+
+use crate::error::MetadataError;
+use crate::Metadata;
+use scale_decode::DecodeAsType;
+
+/// gives you access to all `CustomValueMetadata` entries that are available as custom types on the metadata.
+pub struct CustomTypes {
+    metadata: Metadata,
+}
```

Comment:
Let's mirror constants etc and call this something like `CustomValuesClient` and have it store `Client` in it.

That way, we are consistent with the other "top level" APIs, but also give ourselves room to add eg codegen or whatever to aceess and decode custom types (which I hadn't thought about to be honest, but Alex bought it up and it seems like a reasonable sort of thing to want here, so that people can easily get hold of concrete decoded values).

Now the more contentious bit:

For the same reason (having concrete types/codegen), I wonder whether we should make the API look like the others also with respect to passing in some payload to access the value (ie copy the approach taken in the `constants` APIs)? That way we can have dynamic or static "queries" to access the value much like other APIS.

Pros of doing this:
- Consistent with other APIs in subxt
- Can decode to static values, or `DecodedValueThunk`'s just like constants

Cons:
- More overkill in terms of code to write
- `DecodedValueThunk` copies the bytes when it decodes which is less performant.

I think on balance it's probably worth it to be consistent and give room for dynamic or static decoding. What do you think?

Reply:
I think we should make a CustomValuesClient that can only query values dynamically. I do not see a point in giving static queries to it, because we could just as well expose the static values in the codegen. 
For example if there is a custom type `Foo` with the structure `struct Foo {a: u8, b: bool}` that has an associated value `vec![3,0]`, we could probably generate:
```
struct Foo {a: u8, b: bool}
impl Default for Foo{
    fn default() -> Foo{
        const data: [u8; 2] = [3,0];
        Foo::decode(&mut &data[..]).expect("should work, otherwise Foo value is broken.")
    } 
}
```
This will make it much less confusing for users to use. Sadly we cannot make the Foo default value constant directly, because it might contains vecs and other types with heap allocations.
What do you think about that? Maybe we can also go the same path for constants.